proposals/context.md

```diff
@@ -0,0 +1,265 @@
+        callback: (coolThing) => {
+            this.myCoolThing = coolThing; // do something with value
+        }
+    )
+);
+```
+
+If a provider listening for this event can provide the requested context it will invoke the callback passed in the payload of the event. The element can then do whatever it wishes with this value.
+
+It may also be the case that a provider can retain a reference to this callback, and can then invoke the callback multiple times. In this case providers should pass a dispose function as a second argument to the callback to allow consumers to inform the provider that it should no longer update the element, and should dispose of the callback.
+
+As a convenience, and a hint for providers, an element may also provide a once boolean on the event detail to indicate that it is not interested in receiving updates to the value. If this behavior is essential to the correct operation of the consumer, then they should be implemented defensively as there is no guarantee that providers will honor this agreement. An example is provided below:
+
+```js
+this.dispatchEvent(
+  new RequestContextEvent(
+    "cool-thing-we-want-once",
+    (coolThing, dipose) => {
+      // if we were given a disposer, this provider is likely to send us updates
+      if (dispose) {
+        // so dispose immediately
+        dispose();
+      }
+      // guard against multiple assignment in case of bad actor providers
+      if (!this.myCoolThing) {
+        this.myCoolThing = coolThing; // do something with value
+      }
+    },
+    true // we only want the event once
+  )
+);
+```
+
+It is recommended that custom elements which participate in the Context API should fire their context-request events in their connectedCallback handler. Likewise in their disconnectedCallback they should invoke any dispose functions they have received.
+
+A more complete example is as follows:
+
+```js
+class SimpleElement extends HTMLElement {
+  connectedCallback() {
+    this.dispatchEvent(
+      new RequestContextEvent("logger", (value, dispose) => {
+        // protect against changing providers
+        if (dispose && dispose !== this.loggerDisposer) {
+          this.dispose();
+        }
+        this.logger = value;
+        this.loggerDisposer = dispose;
+      })
+    );
+  }
+  disconnectedCallback() {
+    if (this.loggerDisposer) {
+      this.loggerDisposer();
+    }
+    this.loggerDisposer = undefined;
+    this.logger = undefined;
+  }
+}
+```
+
+# Open Questions / Previous considerations
+
+## Could we use Promises instead of callbacks?
+
+Many have commented that Promises could be used instead of callback functions. The major drawback of promises is that they do not allow multiple-resolution. Therefore we would not be able to handle cases where a requested value changes over time. This is capability that we see in the React Context API and believe is valuable for a variety of use cases.
```

Comment:
Promises can also not be resolved synchronously

Reply:
Expanded on this a little in bccaeac